File: haplovis_front/src/types/layout.ts
export interface Position {
  x: number;
  y: number;
}

export interface Bounds {
  xl: number;
  xr: number;
}

export interface Densities {
  densities: number[];
  down_sample_factor: number;
}

export interface RectangleRange {
  lu: Position;
  rd: Position;
}

export interface LayoutNode {
  segment_id: string;
  position: Position;
}

export type Layout = Record<string, LayoutNode>;

export interface GraphLayoutState {
  viewport: RectangleRange;
  zoom: number;
  pan: Position;
  bufferSize: number;
  extent: Bounds;
  firstGraphRender: boolean;
}

File: src/haplovis/schemas/layout.py
from typing import Dict, List, Tuple
from pydantic import BaseModel, dataclasses


@dataclasses.dataclass
class Position:
    x: float
    y: float


@dataclasses.dataclass
class Bounds:
    xl: float
    xr: float


LayoutCLI = Dict[str, Tuple[Position, Bounds]]


@dataclasses.dataclass
class RectangleRange:
    lu: Position
    rd: Position


class Densities(BaseModel):
    down_sample_factor: int
    densities: List[int]


@dataclasses.dataclass
class LayoutNode:
    segment_id: str
    position: Position


Layout = Dict[str, LayoutNode]
